dtfs-central-api/src/services/utilisation-report-data-validator/index.ts

```diff
@@ -1,137 +1,26 @@
-    { header: UTILISATION_REPORT_HEADERS.FEES_PAID_IN_PERIOD, errorGenerator: generateFeesPaidForThePeriodError },
-    {
-      header: UTILISATION_REPORT_HEADERS.FEES_PAID_IN_PERIOD_CURRENCY,
-      errorGenerator: generateFeesPaidForThePeriodCurrencyError,
-    },
-  ];
-
-  const rowValidations = [
-    generateTotalFeesAccruedCurrencyError,
-    generateTotalFeesAccruedExchangeRateError,
-    generatePaymentCurrencyError,
-    generatePaymentExchangeRateError,
-  ];
-
-  const errors = await Promise.all(
-    csvData.map(async (csvRow) => {
-      const cellErrorPromises = cellValidations.map(({ header, errorGenerator }) => {
-        if (!availableHeaders.includes(header)) {
-          return null;
-        }
-
-        return errorGenerator(csvRow[header], csvRow.exporter?.value);
-      });
-
-      const rowErrorPromises = rowValidations.map((errorGenerator) => errorGenerator(csvRow));
-
-      return Promise.all([...cellErrorPromises, ...rowErrorPromises]);
-    }),
-  );
-
-  // Filter out null errors so that we don't return any null elements in the response array
-  const errorsWithNullValuesFilteredOut = errors.flat().filter((error): error is UtilisationReportDataValidationError => error !== null);
-
-  return errorsWithNullValuesFilteredOut;
-};
+import { UtilisationReportDataValidationError, UtilisationReportCsvRowData } from '@ukef/dtfs2-common';
+import validateHeaders from './utilisation-report-cell-validators/validate-csv-data/headers';
+import validateCells from './utilisation-report-cell-validators/validate-csv-data/cells';
+import validateRows from './utilisation-report-cell-validators/validate-csv-data/rows';
 
 /**
  * Validate the utilisation report csv data
+ * validates the headers
+ * Generates errors for missing headers
+ * Generates an array of available headers
+ * Validates the cell data for each row based on the available headers
+ * Validates the full csv sheet for errors - where rows with same facility ids should have certain fields matching
  * @param csvData - The data from the utilisation report csv
  * @returns An array of errors pertaining to the report if there are any
  */
 export const validateUtilisationReportCsvData = async (csvData: UtilisationReportCsvRowData[]): Promise<UtilisationReportDataValidationError[]> => {
-  const { missingHeaderErrors, availableHeaders } = validateUtilisationReportCsvHeaders(csvData[0]);
+  const { missingHeaderErrors, availableHeaders } = validateHeaders(csvData[0]);
+
+  const cellValidationErrors = await validateCells(csvData, availableHeaders);
 
-  const dataValidationErrors = await validateUtilisationReportCsvCellData(csvData, availableHeaders);
+  const rowValidationErrors = validateRows(csvData);
 
-  const validationErrors = missingHeaderErrors.concat(dataValidationErrors);
+  const validationErrors = [...(missingHeaderErrors ?? []), ...(cellValidationErrors ?? []), ...(rowValidationErrors ?? [])];
```

Comment:
Do we really need the `?? []`, it looks like `missingHeaderErrors` and I assume the others will return an empty array?

Reply:
removed